# AVALIAÇÃO 1 - PROGRAMAÇÃO EM FINANÇAS
#### Professor: Julio Russo
#### Disciplina: Programação em Finanças
#### Turma: 2025.1
#### Alunos: Lucas dos Santos Marques, Luiz Fernando M. Meirinho , Victor A. S. de Souza e Victor Flávio P. Dornelos
#### Data: 16/04/2025


# Introdução


Este projeto tem como objetivo responder às questões propostas na Avaliação 1 da disciplina **Programação em Finanças**. Para isso, o presente notebook reúne tanto a implementação dos códigos em Python quanto as análises correspondentes realizadas a partir dos resultados.

O desenvolvimento está sendo feito no **Visual Studio Code (VS Code)**, mas o código pode ser executado em qualquer IDE compatível com Python. Todas as bibliotecas utilizadas estão listadas no arquivo `requirements.txt`, e o projeto foi desenvolvido utilizando a versão **3.12.0 do Python**.

In [ ]:
pip install -r requirements.txt

In [19]:
# Bibliotecas necessárias
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
from IPython.display import display, Markdown
from tabulate import tabulate

# Questão 1:

### 1) CDBs e Conversão de Taxas

Buscar ao menos **3 CDBs de bancos de sua preferência** que declarem **taxas ao ano** e transformá-las em **taxas capitalizadas mensalmente**, programando esse processo em Python.

Além disso, declarar qual será o **retorno líquido dos investimentos, já descontados impostos e taxas**, para cada uma das opções analisadas.


#### **Premissas Gerais**
- **Valor do Aporte Inicial:** R$10.000,00  
- **Período de Investimento:** 03/04/2024 a 30/03/2026 (24 meses)  
- **Plataforma:** Nubank – Perfil Cliente Normal  
- **Imposto de Renda (IR):** 17,5%  
- **Taxa de Administração:** 0% (isenta)

---

#### **CDBs Selecionados**

| Banco                | Taxa Nominal Anual (a.a) | Alíquota de IR (%) | Taxa de Administração (%) |
|----------------------|---------------------------|---------------------|----------------------------|
| **Banco Original**   | 15,08%                    | 17,5%               | 0%                         |
| **Banco ABC Brasil** | 14,64%                    | 17,5%               | 0%                         |
| **Banco Caixa Geral**| 14,62%                    | 17,5%               | 0%                         |


In [20]:
def analisar_cdbs_evolucao(cdbs, aporte_inicial, meses, ir_percentual):
    registros = []
    resultados_finais = []

    print("🔍 Análise de CDBs\n")

    for banco, taxa_aa in cdbs.items():
        # Taxa mensal composta
        taxa_am = (1 + taxa_aa) ** (1 / 12) - 1

        for t in range(1, meses + 1):
            montante = aporte_inicial * (1 + taxa_am) ** t
            juros = montante - aporte_inicial
            retorno_liquido = juros * (1 - ir_percentual)

            registros.append({
                "Mês": t,
                "Banco": banco,
                "Retorno Líquido (R$)": retorno_liquido
            })

        retorno_final = registros[-1]["Retorno Líquido (R$)"]
        resultados_finais.append({
            "Banco": banco,
            "Taxa Mensal (%)": round(taxa_am * 100, 4),
            "Retorno Líquido Final (R$)": round(retorno_final, 2)
        })

        # Print dos resultados do banco
        print(f"🏦 {banco}")
        print(f" - Taxa mensal: {taxa_am * 100:.4f}%")
        print(f" - Retorno líquido após {meses} meses: R$ {retorno_final:,.2f}\n")

    # DataFrame com evolução
    df_evolucao = pd.DataFrame(registros)

    # Gráfico
    fig = px.line(df_evolucao,
                  x="Mês",
                  y="Retorno Líquido (R$)",
                  color="Banco",
                  title="📈 Evolução do Retorno Líquido dos CDBs",
                  markers=True)
    
    fig.update_layout(height=500)
    fig.show()

    # Ranking final
    df_ranking = pd.DataFrame(resultados_finais)
    df_ranking = df_ranking.sort_values(by="Retorno Líquido Final (R$)", ascending=False).reset_index(drop=True)

    print("🏆 Ranking dos CDBs (maior para menor retorno líquido):\n")
    print(tabulate(df_ranking, headers='keys', tablefmt='grid', showindex=False))

In [21]:
cdbs = {
    "Banco Original": 0.1508,
    "Banco ABC Brasil": 0.1464,
    "Banco Caixa Geral": 0.1462
}

analisar_cdbs_evolucao(cdbs, aporte_inicial=10000, meses=24, ir_percentual=0.175)


🔍 Análise de CDBs

🏦 Banco Original
 - Taxa mensal: 1.1774%
 - Retorno líquido após 24 meses: R$ 2,675.81

🏦 Banco ABC Brasil
 - Taxa mensal: 1.1451%
 - Retorno líquido após 24 meses: R$ 2,592.42

🏦 Banco Caixa Geral
 - Taxa mensal: 1.1436%
 - Retorno líquido após 24 meses: R$ 2,588.64



🏆 Ranking dos CDBs (maior para menor retorno líquido):

+-------------------+-------------------+------------------------------+
| Banco             |   Taxa Mensal (%) |   Retorno Líquido Final (R$) |
+===================+===================+==============================+
| Banco Original    |            1.1774 |                      2675.81 |
+-------------------+-------------------+------------------------------+
| Banco ABC Brasil  |            1.1451 |                      2592.42 |
+-------------------+-------------------+------------------------------+
| Banco Caixa Geral |            1.1436 |                      2588.64 |
+-------------------+-------------------+------------------------------+


### 📊 Análise Comparativa dos CDBs

Devido ao horizonte de investimento relativamente curto — apenas dois anos — as diferenças de rentabilidade mensal entre os títulos avaliados tornam-se menos expressivas. Isso se deve, em parte, ao baixo volume aplicado, pois percentuais de retorno ganham maior relevância à medida que os valores envolvidos se tornam mais elevados, especialmente quando ultrapassam a casa dos milhões.

Para o cálculo da conversão a taxa de juros anual para mensal, foi utilizado a fórmula de equivalência de taxas de diferentes frequências. Posteriormente, foi feito apenas uma conta simples de juros compostos para saber quanto renderia o aporte ao final do contrato, visto que é uma única aplicação. O retorno líquido final é descontando o valor inicial aplicado e os impostos.

Outro ponto importante é que nenhum dos CDBs analisados cobra taxa de administração, o que contribui para um resultado mais transparente na comparação entre eles e todos são capitalizados mensalmente.

Dessa forma, é possível destacar que o **Banco Original** oferece uma rentabilidade significativamente superior em relação aos demais. O **CDB do Banco ABC** também apresenta desempenho ligeiramente melhor que o da **Caixa Geral**, embora a diferença entre ambos seja pouco representativa. Onde, os retornos líquidos podem ser visto abaixo.

#### 🏆 Ranking dos CDBs (Maior para Menor Retorno Líquido)

| Banco             | Taxa Mensal (%) | Retorno Líquido Final (R$) |
|-------------------|------------------|------------------------------|
| **Banco Original**    | **1,1774**         | **R$ 2.675,81**                  |
| Banco ABC Brasil  | 1,1451         | R$ 2.592,42                  |
| Banco Caixa Geral | 1,1436         | R$ 2.588,64                  |


# Questão 2:

### 2) Comparação de Retorno: Tesouro IPCA+ vs. Juros Pré-fixados

Programar em Python uma simulação para comparar o **retorno líquido** entre o **Tesouro IPCA+** e um título com **juros pré-fixados**, considerando **tempos de vencimentos equivalentes**.

O valor do investimento monetário fica a **critério do grupo simular**.

O objetivo é verificar **qual dos dois investimentos se torna mais vantajoso após os descontos de impostos e taxas**.


## Premissas

- **Aplicação única**: R$10.000
- **Inflação estimada (IPCA)**: 4% ao ano
- **Imposto de Renda (IR)**: a definir conforme prazo (≥ 720 dias = 15%)
- **Prazo da aplicação**: até 01/01/2035
- **Data de aplicação**: 03/02/2025
- **Taxa de administração**: 0,2% ao ano
- **Pagamento de juros**: considerado apenas no vencimento (para simplificação)

---

## Títulos Públicos Analisados

| Título                          | Tipo de Rentabilidade       | Taxa de Juros              | Vencimento   | Fonte (Consulta)    |
|--------------------------------|-----------------------------|----------------------------|--------------|----------------------|
| Tesouro IPCA+ 2035             | Pós-fixado (IPCA + real)    | IPCA + 7,49% a.a           | 01/01/2035   | Tesouro Direto (03/02/2025) |
| Tesouro Prefixado 2035         | Pré-fixado                  | 14,76% a.a                 | 01/01/2035   | Tesouro Direto (03/02/2025) |

> **Observação**: Para fins de simplificação, considerou-se que o pagamento dos juros ocorre integralmente no vencimento final (sem cupons semestrais).



In [22]:
# Dados da aplicação
valor_investido = 10000  # Valor da aplicação única
prazo_meses = 119
prazo_dias = prazo_meses * 30  # Considerando mês comercial de 30 dias
dias = np.arange(0, prazo_dias + 1)

# Taxas anuais
infla_1 = 0.04   # Cenário 1: 4%
infla_2 = 0.055  # Cenário 2: 5,5%
taxa_ipca_mais_real = 0.0749
taxa_prefixado_aa = 0.1476  # 14,76% a.a.

# Parâmetros
ir = 0.15            # Imposto de renda para aplicações longas
taxa_adm_aa = 0.002  # Taxa de administração anual

# Função para taxa efetiva diária
def taxa_dia(taxa_anual):
    return (1 + taxa_anual) ** (1 / 360) - 1

# Cálculo para os dois cenários de IPCA
cenarios_ipca = {'4%': infla_1, '5,5%': infla_2}
resultados_ipca = {}

for nome, ipca in cenarios_ipca.items():
    taxa_ipca_mais_aa = (1 + ipca) * (1 + taxa_ipca_mais_real) - 1
    taxa_dia_ipca = taxa_dia(taxa_ipca_mais_aa)
    taxa_adm_dia = taxa_dia(taxa_adm_aa)
    taxa_liq_dia_ipca = taxa_dia_ipca - taxa_adm_dia

    montante_bruto = valor_investido * (1 + taxa_liq_dia_ipca) ** dias
    lucro = montante_bruto[-1] - valor_investido
    valor_final = montante_bruto[-1] - lucro * ir
    retorno = valor_final - valor_investido

    resultados_ipca[nome] = {
        "taxa_liquida": taxa_liq_dia_ipca * 360,
        "valor_final": valor_final,
        "retorno": retorno,
        "montante_mensal": montante_bruto[::30] * (1 - ir) + valor_investido * ir
    }

# Tesouro Prefixado (único cenário)
taxa_dia_prefixado = taxa_dia(taxa_prefixado_aa)
taxa_adm_dia = taxa_dia(taxa_adm_aa)
taxa_liq_dia_prefixado = taxa_dia_prefixado - taxa_adm_dia

montante_prefixado_bruto = valor_investido * (1 + taxa_liq_dia_prefixado) ** dias
lucro_prefixado = montante_prefixado_bruto[-1] - valor_investido
valor_final_prefixado = montante_prefixado_bruto[-1] - lucro_prefixado * ir
retorno_prefixado = valor_final_prefixado - valor_investido
valores_prefixado_mensal = montante_prefixado_bruto[::30] * (1 - ir) + valor_investido * ir

In [23]:
# Gráfico
meses = np.arange(1, len(valores_prefixado_mensal) + 1)
df_plot = pd.DataFrame({
    'Meses': meses,
    'Tesouro Prefixado 2035': valores_prefixado_mensal,
    'Tesouro IPCA+ 2035 (IPCA 4%)': resultados_ipca['4%']['montante_mensal'],
    'Tesouro IPCA+ 2035 (IPCA 5,5%)': resultados_ipca['5,5%']['montante_mensal']
})

df_melted = df_plot.melt(id_vars='Meses', var_name='Investimento', value_name='Valor')

fig = px.line(df_melted, x='Meses', y='Valor', color='Investimento',
              title='Evolução do Valor Investido ao Longo do Tempo (Cenários IPCA)',
              labels={'Valor': 'Valor Acumulado (R$)', 'Meses': 'Tempo (meses)'})
fig.show()

# Tabela resumo
resumo = pd.DataFrame([
    {
        "Investimento": "Tesouro Prefixado 2035",
        "Taxa Líquida Aproximada (%)": round(taxa_liq_dia_prefixado * 100 * 360, 4),
        "Valor Final (R$)": round(valor_final_prefixado, 2),
        "Retorno Líquido (R$)": round(retorno_prefixado, 2)
    },
    {
        "Investimento": "Tesouro IPCA+ 2035 (IPCA 4%)",
        "Taxa Líquida Aproximada (%)": round(resultados_ipca['4%']['taxa_liquida'] * 100, 4),
        "Valor Final (R$)": round(resultados_ipca['4%']['valor_final'], 2),
        "Retorno Líquido (R$)": round(resultados_ipca['4%']['retorno'], 2)
    },
    {
        "Investimento": "Tesouro IPCA+ 2035 (IPCA 5,5%)",
        "Taxa Líquida Aproximada (%)": round(resultados_ipca['5,5%']['taxa_liquida'] * 100, 4),
        "Valor Final (R$)": round(resultados_ipca['5,5%']['valor_final'], 2),
        "Retorno Líquido (R$)": round(resultados_ipca['5,5%']['retorno'], 2)
    }
])
resumo["Ranking"] = resumo["Valor Final (R$)"].rank(ascending=False).astype(int)

resumo

Investimento  Taxa Líquida Aproximada (%)  \
0          Tesouro Prefixado 2035                      13.5701   
1    Tesouro IPCA+ 2035 (IPCA 4%)                      10.9468   
2  Tesouro IPCA+ 2035 (IPCA 5,5%)                      12.3792   

   Valor Final (R$)  Retorno Líquido (R$)  Ranking  
0          34139.25              24139.25        1  
1          26665.02              16665.02        3  
2          30504.84              20504.84        2

### 📊 Explicação dos cenários:

Escolhemos dois cenários para nossa análise desses titulos públicos

## Cenário 1

Inflação escolhida no valor de 4% ao ano de média para os próximos 10 anos, correspondentes ao período de investimento

Entendemos como um cenário otimista, visto que no histórico desde o plano real, existe uma dificuldade do banco central em manter taxas de inflação a patamares menores que 4%

Além do cenário de crise fiscal brasileiro, que dificulta a convergência da meta de inflação de 3%

## Cenário 2

Inflação escolhida no valor de 5,5% ao ano de média para os próximos 10 anos, correspondentes ao período de investimento

Entendemos como um cenário mais realista, porque se aproxima da média histórica dos últimos 5 anos.

Além de ser mais compatível com um cenário de dificuldade de ajuste fiscal


### 📊 Resultados:

De acordo com os dois cenários acima descritos, chegamos aos seguintes resultados

Tesouro prefixado teve um retorno líquido de 24.139,25 reais

Tesouro Ipca+2035 no cenário 1 teve um retorno líquido de 16.665,02 reais

Tesouro Ipca+2035 no cenário 2 teve um retorno líquido de 20.504,82 reais


### 📊 Análise:

De acordo com os resultados acima, o titulo que possui a melhor rentabilidade é o Tesouro Prefixado 2035

Isso significa que nesse momento o mercado está precificando nesse titulo uma inflação futura maior que em ambos os cenários realizados em nossos testes

Dessa forma, o investidor pode supor que a inflação futura não seja tão elevada, se aproximando de cenários mais realistas, como o cenário 2

Nesse caso, o investidor deve realizar o seu investimento no titulo prefixado.

Apenas ressaltando que caso a inflação cresça a patamares muito elevados, muito acima do histórico do plano real, o tesouro IPCA+ se tornará mais rentável


# Questão 3:

### 3) Programar em Python o quanto de aporte financeiro é necessário para alcançar em 7 anos, 10 anos e 12 anos:

- A) R$30.000  
- B) R$50.000  
- C) R$100.000  

Considerando uma **taxa de juros real efetiva de 7% ao ano** em todos esses cenários.


### Premissas:
 Sendo utilizado a fórmula de valor futuro de uma série uniforme postecipada capitalizada mensalmente, sendo assim foi feito a conversão da taxa de juros anual para mensal. 
 
 Consideramos um aporte financeiro mensal, pois grande parte da população recebe sua renda mensalmente e destina parte dela para investimento e ao final obtemos o valor da aplicação total.

 Outro detalhe, pelo fato do enunciado tratar como taxa de juros real efetiva, não será incluso taxa de impostos e/ou de administração porque pode variar com o tipo de investimento.

In [24]:
def gerar_dados(montante_desejado, anos, taxa_juros_anual, aliquota_ir=0):
    # Aplica o imposto na taxa de juros
    taxa_juros_anual_liquida = taxa_juros_anual * (1 - aliquota_ir)

    # Converte a taxa anual líquida para taxa efetiva mensal
    i = (1 + taxa_juros_anual_liquida) ** (1 / 12) - 1
    n = anos * 12

    # Cálculo da parcela mensal (R)
    R = montante_desejado * i / ((1 + i) ** n - 1)

    # Montantes acumulados mês a mês
    meses = np.arange(1, n + 1)
    montantes = R * ((1 + i) ** meses - 1) / i

    # DataFrame
    df = pd.DataFrame({
        'Mês': meses,
        'Montante Acumulado (R$)': montantes,
        'Prazo (anos)': f'{anos} anos'
    })

    return R, df

In [25]:
def calcular_para_varios_anos_plotly(montante_desejado, anos_lista, taxa_juros_anual=0.07, aliquota_ir=0):
    resultados = {}  # Dicionário para guardar o valor da parcela mensal para cada prazo
    dfs = []  # Lista para acumular os DataFrames de cada prazo

    for anos in anos_lista:
        R, df = gerar_dados(montante_desejado, anos, taxa_juros_anual, aliquota_ir)  # Gera os dados para cada prazo
        resultados[anos] = R  
        dfs.append(df) 

        # Mostra a parcela mensal e o total aportado para o prazo atual
        display(Markdown(f"**{anos} anos** → parcela mensal R$ {R:.2f} e aporte total R$ {R * anos * 12:.2f}"))

        # Gráfico da evolução do montante acumulado para o prazo atual
        fig = px.line(df, x='Mês', y='Montante Acumulado (R$)',
                      title=f'Evolução do Montante - {anos} anos',
                      markers=True)
        fig.update_layout(template='plotly_white')
        fig.show()

    # Junta todos os DataFrames para o gráfico comparativo
    df_total = pd.concat(dfs, ignore_index=True)

    # Gráfico comparativo da evolução do montante para todos os prazos
    fig_total = px.line(df_total, x='Mês', y='Montante Acumulado (R$)', color='Prazo (anos)',
                        title=f'Evolução do Montante para Meta de R$ {montante_desejado:,.2f} - Todos os Prazos',
                        markers=True)
    fig_total.update_layout(template='plotly_white')
    fig_total.show()

    return resultados 



## A)

In [26]:
calcular_para_varios_anos_plotly(
    montante_desejado=30000,
    anos_lista=[7, 10, 12],
    taxa_juros_anual=0.07,
    aliquota_ir=0 
)


**7 anos** → parcela mensal R$ 280.01 e aporte total R$ 23520.77

**10 anos** → parcela mensal R$ 175.39 e aporte total R$ 21046.26

**12 anos** → parcela mensal R$ 135.46 e aporte total R$ 19506.53

{7: 280.0091587617208, 10: 175.38553851336093, 12: 135.46198770024267}

### 📊 Análise:
Observa-se que o valor de aporte necessário diminui conforme o tempo passa, exemplificando a potencialidade do tempo.

**Quanto maior o tempo, maior o efeito dos juros, e menor o aporte necessário.**

Dessa forma, para alcançar a marca de R$ 30 mil, é necessário:

Aportes mensais de R$ 280 reais e total de R$ 23,5 mil, caso almeje atingir o montante em sete anos.

Aportes mensais de R$ 175.38 reais e total de R$ 21 mil, caso o tempo alvo seja de dez anos.

E aportes mensais de R$ 135.46 e total de R$ 19,5 mil, caso o horizonte temporal seja de doze anos.

## B)

In [27]:
calcular_para_varios_anos_plotly(
    montante_desejado=50000,
    anos_lista=[7, 10, 12],
    taxa_juros_anual=0.07,
    aliquota_ir=0 
)


**7 anos** → parcela mensal R$ 466.68 e aporte total R$ 39201.28

**10 anos** → parcela mensal R$ 292.31 e aporte total R$ 35077.11

**12 anos** → parcela mensal R$ 225.77 e aporte total R$ 32510.88

{7: 466.68193126953474, 10: 292.3092308556016, 12: 225.76997950040445}

### 📊  Análise:


**Quanto maior o tempo, maior o efeito dos juros sobre o montante, e menor a necessidade de aporte**

Nesse caso, o objetivo é alcançar R$50 mil reais, que aos juros de 7% ao ano necessitará:

De aportes mensais de R$466.68 e total de R$ 39,2 mil em sete anos.

Aportes mensais de R$292.31 e total de R$ 35 mil em dez anos.

E aportes mensais de R$225.77 e total de R$ 32,5 mil em dozes anos

## C)

In [28]:
calcular_para_varios_anos_plotly(
    montante_desejado=100000,
    anos_lista=[7, 10, 12],
    taxa_juros_anual=0.07,
    aliquota_ir=0 
)


**7 anos** → parcela mensal R$ 933.36 e aporte total R$ 78402.56

**10 anos** → parcela mensal R$ 584.62 e aporte total R$ 70154.22

**12 anos** → parcela mensal R$ 451.54 e aporte total R$ 65021.75

{7: 933.3638625390695, 10: 584.6184617112032, 12: 451.5399590008089}

### 📊 Análise:


**Quanto maior o tempo, maior o efeito dos juros sobre o montante, e menor a necessidade de aporte**

Nesse caso, o objetivo é alcançar R$50 mil reais, que aos juros de 7% ao ano necessitará:

De aportes mensais de R$ 933.36 e total de R$ 78,4 mil em sete anos.

Aportes mensais de R$ 584.62 e total de R$ 70,2 mil em dez anos.

E aportes mensais de R$ 451.54 e total de R$ 65 mil em dozes anos

## 🏆 Conclusão:

Quanto maior o montante futuro desejado, maior será o valor do aporte financeiro.

Quanto maior o juros real efetivo, menor será o valor do aporte financeiro.

Quanto maior o tempo de espera para obter o montante, menor será o valor do aporte financeiro.

Quanto maior as taxas de administração e/ou impostos, menor será o valor do aporte financeiro.

Quanto maior a frequência da capitalização e/ou do aporte financeiro, menor será o valor do aporte financeiro.